components/uniset/src/builder.rs

```diff
@@ -0,0 +1,755 @@
+    ///
+    /// Performs in `O(B + S)`, where `B` is the number of endpoints in the Builder, and `S` is the number
+    /// of endpoints in the argument.
+    fn complement_list(&mut self, set: Vec<u32>) {
+        let mut res: Vec<u32> = vec![]; // not the biggest fan of having to allocate new memory
+        let mut ai = self.intervals.iter();
+        let mut bi = set.iter();
+        let mut a = ai.next();
+        let mut b = bi.next();
+        while let (Some(c), Some(d)) = (a, b) {
+            match c.cmp(&d) {
+                Ordering::Less => {
+                    res.push(*c);
+                    a = ai.next();
+                }
+                Ordering::Greater => {
+                    res.push(*d);
+                    b = bi.next();
+                }
+                Ordering::Equal => {
+                    a = ai.next();
+                    b = bi.next();
+                }
+            }
+        }
+        if let Some(c) = a {
+            res.push(*c)
+        }
+        if let Some(d) = b {
+            res.push(*d)
+        }
+        res.extend(ai);
+        res.extend(bi);
+        self.intervals = res;
+    }
+
+    /// Computes the complement of the builder, inverting the builder (any elements in the builder are removed,
+    /// while any elements not in the builder are added)
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::{UnicodeSetBuilder, UnicodeSet};
+    /// let mut builder = UnicodeSetBuilder::new();
+    /// let set = UnicodeSet::from_inversion_list(vec![0, 65, 70, (std::char::MAX as u32) + 1]).unwrap();
+    /// builder.add_set(&set);
+    /// builder.complement();
+    /// let check = builder.build();
+    /// assert_eq!(check.iter().next(), Some('A'));
+    /// ```
+    pub fn complement(&mut self) {
+        if !self.intervals.is_empty() {
+            if self.intervals[0] == 0 {
+                self.intervals.drain(0..1);
+            } else {
+                self.intervals.insert(0, 0);
+            }
+            if self.intervals.last() == Some(&((char::MAX as u32) + 1)) {
+                self.intervals.pop();
+            } else {
+                self.intervals.push((char::MAX as u32) + 1);
+            }
+        }
+    }
+
+    /// Complements the character in the builder, adding it if not in the builder, and removing it otherwise.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSetBuilder;
+    /// let mut builder = UnicodeSetBuilder::new();
+    /// builder.add_range(&('A'..'D'));
+    /// builder.complement_char('A');
+    /// builder.complement_char('E');
+    /// let check = builder.build();
+    /// assert!(check.contains('E'));
+    /// assert!(!check.contains('A'));
+    /// ```
+    pub fn complement_char(&mut self, c: char) {
+        let code_point = c as u32;
+        let to_complement = vec![code_point, code_point + 1];
```

Comment:
Make this a 2-element array and then pass it as a slice to complement_list.

```rust
let code_point = c as u32;
let to_complement = [code_point, code_point + 1];
self.complement_list(&to_complement);
```

Reply:
agreed and changed